### PR TITLE
dms_replication_subnet_group 

### PR DESCRIPTION
**Pull request by ruimoreira**
##### SUMMARY
New module for dms_replication_subnet_group

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME
dms_replication_subnet_group

### CODE REVIEW

File: lib/ansible/modules/cloud/amazon/dms_replication_subnet_group.py [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+def create_module_params(module):
+    """
+    Reads the module parameters and returns a dict
+    :return: dict
+    """
+    instance_parameters = dict(
+        # ReplicationSubnetGroupIdentifier gets translated to lower case anyway by the API
+        ReplicationSubnetGroupIdentifier=module.params.get('identifier').lower(),
+        ReplicationSubnetGroupDescription=module.params.get('description'),
+        SubnetIds=module.params.get('subnetids'),
```
Comment (jillr) [inline lib/ansible/modules/cloud/amazon/dms_replication_subnet_group.py]:
```suggestion
        SubnetIds=module.params.get('subnet_ids'),
```
and here as well

File: lib/ansible/modules/cloud/amazon/dms_replication_subnet_group.py [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+            - The name for the replication subnet group.
+              This value is stored as a lowercase string.
+              Must contain no more than 255 alphanumeric characters,
+              periods, spaces, underscores, or hyphens. Must not be "default".
+        type: str
+    description:
+        description:
+            - The description for the subnet group.
+        type: str
+    subnetids:
```
Comment (jillr) [inline lib/ansible/modules/cloud/amazon/dms_replication_subnet_group.py]:
```suggestion
    subnet_ids:
```
Snake casing would help with readability of this option.

File: lib/ansible/modules/cloud/amazon/dms_replication_subnet_group.py [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+        subnetids=dict(type='list', elements='str', required=True),
+    )
+    module = AnsibleAWSModule(
+        argument_spec=argument_spec,
+        supports_check_mode=True
+    )
+    exit_message = None
+    changed = False
+    if not HAS_BOTO3:
+
```
Comment (jillr) [inline lib/ansible/modules/cloud/amazon/dms_replication_subnet_group.py]:
```suggestion

``` 
nit, but please remove this extra line break.

File: test/integration/targets/dms_replication_subnet_group/tasks/main.yml [hunk 4]
```diff
@@ -0,0 +1,144 @@
+---
+
+- name: set connection information for all tasks
+  set_fact:
+    aws_connection_info: &aws_connection_info
```
Comment (jillr) [inline test/integration/targets/dms_replication_subnet_group/tasks/main.yml:5]: You'll need to use the connection info in each of the test tasks: 
    <<: *aws_connection_info

File: test/integration/targets/dms_replication_subnet_group/tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+    - name: create subnet2
+      ec2_vpc_subnet:
+        state: present
+        vpc_id:  "{{ testing_vpc.vpc.id }}"
+        cidr: 10.22.32.32/28
+        az: eu-west-1c
+      register: subnet2
+
+    - name: create replication subnet subnet group
```
Comment (jillr) [inline test/integration/targets/dms_replication_subnet_group/tasks/main.yml]:
```suggestion
    - name: create replication subnet group
```

File: test/integration/targets/dms_replication_subnet_group/tasks/main.yml [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+
+- name: set connection information for all tasks
+  set_fact:
+    aws_connection_info: &aws_connection_info
+      aws_access_key: "{{ aws_access_key }}"
+      aws_secret_key: "{{ aws_secret_key }}"
+      region: "{{ aws_region }}"
+    resource_prefix: "test_dms_sg"
+    dms_sg_identifier: "{{ resource_prefix }}-dms"
+  no_log: no
```
Comment (jillr) [inline test/integration/targets/dms_replication_subnet_group/tasks/main.yml]:
```suggestion
  no_log: yes
```

**Review by jillr — CHANGES_REQUESTED** [review]
Thanks for this module @ruimoreira . The tests are not passing for me locally but I believe it's unrelated to this change, I will look at them again later this week.

**Comment by ruimoreira** [comment]
Hello @jillr 

I will make these changes shortly, thank you for your help.

**Comment by ruimoreira** [comment]
Hello @jillr 

Your changes have been committed. Thank you for your help.

File: lib/ansible/modules/cloud/amazon/dms_replication_subnet_group.py [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+    # need to sanitize values that get retured from the API
+    if 'VpcId' in param_described.keys():
+        param_described.pop('VpcId')
+    if 'SubnetGroupStatus' in param_described.keys():
+        param_described.pop('SubnetGroupStatus')
+    for paramname in modparams.keys():
+        if paramname in param_described.keys() and \
+                param_described.get(paramname) == modparams[paramname]:
+            pass
+        elif paramname == 'subnet_ids':
```
Comment (jillr) [inline lib/ansible/modules/cloud/amazon/dms_replication_subnet_group.py]:
```suggestion
        elif paramname == 'SubnetIds':
```

File: lib/ansible/modules/cloud/amazon/dms_replication_subnet_group.py [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+def create_module_params(module):
+    """
+    Reads the module parameters and returns a dict
+    :return: dict
+    """
+    instance_parameters = dict(
+        # ReplicationSubnetGroupIdentifier gets translated to lower case anyway by the API
+        ReplicationSubnetGroupIdentifier=module.params.get('identifier').lower(),
+        ReplicationSubnetGroupDescription=module.params.get('description'),
+        subnet_ids=module.params.get('subnet_ids'),
```
Comment (jillr) [inline lib/ansible/modules/cloud/amazon/dms_replication_subnet_group.py]:
```suggestion
        SubnetIds=module.params.get('subnet_ids'),
```
Here this var needs to stay as you had it originally - ansible prefers snake cased but boto typically needs camelcase.  `instance_parameters` will be passed into the boto connection so needs to match what the API expects, both here and later when you access the returned parameters.
https://docs.aws.amazon.com/dms/latest/APIReference/API_CreateReplicationSubnetGroup.html#API_CreateReplicationSubnetGroup_RequestSyntax

File: lib/ansible/modules/cloud/amazon/dms_replication_subnet_group.py [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+        param_described.pop('SubnetGroupStatus')
+    for paramname in modparams.keys():
+        if paramname in param_described.keys() and \
+                param_described.get(paramname) == modparams[paramname]:
+            pass
+        elif paramname == 'subnet_ids':
+            subnets = []
+            for subnet in param_described.get('Subnets'):
+                subnets.append(subnet.get('SubnetIdentifier'))
+            for modulesubnet in modparams['subnet_ids']:
```
Comment (jillr) [inline lib/ansible/modules/cloud/amazon/dms_replication_subnet_group.py]:
```suggestion
            for modulesubnet in modparams['SubnetIds']:
```

File: test/integration/targets/dms_replication_subnet_group/tasks/main.yml [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+        state: present
+        vpc_id:  "{{ testing_vpc.vpc.id }}"
+        cidr: 10.22.32.32/28
+        az: eu-west-1c
+        <<: *aws_connection_info
+      register: subnet2
+
+    - name: create replication subnet group
+      dms_replication_subnet_group:
+        state: present
```
Comment (jillr) [inline test/integration/targets/dms_replication_subnet_group/tasks/main.yml:56]: This action requires dms-vpc-role.
https://docs.aws.amazon.com/dms/latest/userguide/CHAP_Security.APIRole.html
https://forums.aws.amazon.com/thread.jspa?messageID=694553
```
botocore.errorfactory.AccessDeniedFault: An error occurred (AccessDeniedFault) when calling the CreateReplicationSubnetGroup operation: The IAM Role arn:aws:iam::112233445566:ro
le/dms-vpc-role is not configured properly.

fatal: [testhost]: FAILED! => {
    "changed": false,
    "error": {
        "code": "AccessDeniedFault",
        "message": "The IAM Role arn:aws:iam::112233445566:role/dms-vpc-role is not configured properly."
    },
```
Examples of creating a required role in a test can be seen in https://github.com/ansible/ansible/tree/devel/test/integration/targets/aws_codepipeline  or https://github.com/ansible/ansible/tree/devel/test/integration/targets/aws_secret/

**Review by jillr — CHANGES_REQUESTED** [review]

File: test/integration/targets/dms_replication_subnet_group/tasks/main.yml (hunk 10 above)
Comment (ruimoreira) [reply test/integration/targets/dms_replication_subnet_group/tasks/main.yml:56]: Hello , 

this should now work, would you mind trying again please ?

File: test/integration/targets/dms_replication_subnet_group/tasks/main.yml [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+        identifier: "dev-sngroup"
```
Comment (jillr) [inline test/integration/targets/dms_replication_subnet_group/tasks/main.yml]:
```suggestion
        identifier: "{{ dms_sg_identifier }}"
```

File: test/integration/targets/dms_replication_subnet_group/tasks/main.yml [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+        description: "Development Subnet Group asdasdas"
```
Comment (jillr) [inline test/integration/targets/dms_replication_subnet_group/tasks/main.yml]:
```suggestion
        description: "Development Subnet Group"
```
The name and description in this task and the ones in "name: create subnet group no change" are different, so the no-change assert is failing.

File: lib/ansible/modules/cloud/amazon/dms_replication_subnet_group.py (hunk 9 above)
Comment (jillr) [reply lib/ansible/modules/cloud/amazon/dms_replication_subnet_group.py]: This one still needs to be swapped.

**Review by jillr — CHANGES_REQUESTED** [review]
Thanks for those updates, just a couple more comments.

**Comment by ruimoreira** [comment]
Thank you so much for your help in this. 
This should hopefully now pass the tests.

**Review by jillr — APPROVED** [review]
Looks good, thanks for the work on this!

**Pull request merged**